chia/consensus/block_body_validation.py

```diff
@@ -154,24 +154,42 @@ async def validate_block_body(
     if height > constants.INITIAL_FREEZE_PERIOD and constants.NETWORK_TYPE == NetworkType.MAINNET:
         return Err.INITIAL_TRANSACTION_FREEZE, None
     else:
-        # 6. The generator root must be the tree-hash of the generator (or zeroes if no generator)
+        # 6a. The generator root must be the hash of the serialized bytes of
+        #     the generator for this block (or zeroes if no generator)
         if block.transactions_generator is not None:
-            if block.transactions_generator.get_tree_hash() != block.transactions_info.generator_root:
+            if std_hash(bytes(block.transactions_generator)) != block.transactions_info.generator_root:
                 return Err.INVALID_TRANSACTIONS_GENERATOR_ROOT, None
         else:
             if block.transactions_info.generator_root != bytes([0] * 32):
                 return Err.INVALID_TRANSACTIONS_GENERATOR_ROOT, None
 
+        # 6b. The generator_ref_list must be the hash of the serialized bytes of
+        #     the generator ref list for this block (or zeroes if no generator)
+        # 6c. The generator ref list length must be less than or equal to MAX_GENERATOR_REF_LIST_SIZE entries
         if block.transactions_generator_ref_list is not None:
-            if len(bytes(block.transactions_generator_ref_list)) > constants.MAX_GENERATOR_REF_LIST_SIZE:
-                return Err.PRE_SOFT_FORK_MAX_GENERATOR_REF_LIST_SIZE, None
+            if std_hash(bytes(block.transactions_generator)) != block.transactions_info.generator_root:
+                return Err.INVALID_TRANSACTIONS_GENERATOR_REFS_ROOT, None
+            if len(block.transactions_generator_ref_list) > constants.MAX_GENERATOR_REF_LIST_SIZE:
+                return Err.MAX_GENERATOR_REF_LIST_SIZE, None
+        else:
+            if block.transactions_info.generator_refs_root != bytes([0] * 32):
+                return Err.INVALID_TRANSACTIONS_GENERATOR_REFS_ROOT, None
 
         if block.transactions_generator is not None:
-            # Get List of names removed, puzzles hashes for removed coins and conditions crated
+            # The generator must be less than MAX_GENERATOR_SIZE bytes in length
+            if len(block.transactions_generator) <= MAX_GENERATOR_SIZE:
+                return Err.PRE_SOFT_FORK_MAX_GENERATOR_SIZE
+
+            # Get List of names removed, puzzles hashes for removed coins and conditions created
             if cached_cost_result is not None:
                 result: Optional[CostResult] = cached_cost_result
             else:
                 result = calculate_cost_of_program(block.transactions_generator, constants.CLVM_COST_RATIO_CONSTANT)
+            # XXX if result == None, should we add a new error here?
+            # e.g. Err.INVALID_GENERATOR_PROGRAM / FAILED_COST_CALCULATION
+            if result is None:
```

Comment:
This was just an assert before.
If the block is from the blockchain, this should never fail.
If we created the block, it could be an internal error when creating the generator.